### PR TITLE
[SPARK-38378][SQL] Refactoring of the ANTLR grammar definition into separate Parser and Lexer files

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?

Separating the mixed parser grammar defined in `SqlBase.g4` into separate parser and lexer grammars.
* The parser grammar disallows any literal definitions. Thus all literals are replaced with names defined in the new lexer.
* The lexer and parser has to be provided in two files. As ANTLR only allows one grammar per file.

### Why are the changes needed?

This gives us a cleaner separation of parser and lexer in the original grammar. 
It also enables us to use the full power of ANTLR parser and lexer grammars: 
* Access to lexer specific rules: lexer specific rules (e.g. [Lexer mode](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexical-modes)) can be used for new SQL features.
* Ability to reuse lexer rules: we can now use inheritance to have multiple lexers sharing a common set of lexical rules.
* Clear order of tokens: the order the tokens are tokenized by the lexer in the order they appear in the lexer.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests. The refactoring should not break any tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/SQLKeywordSuite.scala [hunk 1]
```diff
@@ -30,11 +30,15 @@ import org.apache.spark.sql.catalyst.util.fileToString
 trait SQLKeywordUtils extends SparkFunSuite with SQLHelper {
 
   val sqlSyntaxDefs = {
-    val sqlBasePath = {
+    val sqlBaseParserPath =
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/SQLKeywordSuite.scala:33]: We don't need to read this file now. All the keywords are defined as lexer rules.

Comment (zhenlineo) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/SQLKeywordSuite.scala:33]: We still have `ansiNonReserved` `nonReserved` defined in parser file. So the test need to include both files.

**Comment by zhenlineo** [comment]
It is a best practice to use ANTLR for a complex grammar. 
With an explicit lexer, we now could: 
1. Access lexer specific rules: lexer specific rules (e.g. [Lexer mode](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexical-modes)) can be used for new SQL features.
2. Reuse lexer rules: we can now use inheritance to have multiple lexers sharing a common set of lexical rules.
3. Define the order of tokens: the order the tokens are tokenized by the lexer in the order they appear in the lexer.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**